src/v/cloud_storage/remote_segment.cc

```diff
@@ -50,6 +50,28 @@
 
 #include <exception>
 
+namespace {
+class bounded_stream final : public ss::data_source_impl {
+public:
+    bounded_stream(ss::input_stream<char>& stream, size_t upto)
+      : _stream{stream}
+      , _upto{upto} {}
+
+    ss::future<ss::temporary_buffer<char>> get() override {
+        const auto buf = co_await _stream.read_up_to(_upto);
+        if (buf.size() < _upto) {
+            _upto -= buf.size();
+        }
+        co_return buf;
+    }
```

Comment:
What prevents _upto from underflowing? Is it possible for us to read more than the stream size in the first call? If so, would it make sense for subsequent calls to check _upto == 0 and return empty? Or is this only expected to be called once?

Reply:
I have switched to `read_exactly`, I think it is more appropriate here, it will only return less than `_upto` bytes in case of EOF, in which case the next call to our `get` will return an empty buffer signalling EOF correctly.